### PR TITLE
Additional checks regarding control state in IK

### PR DESCRIPTION
**Pull request by bli76**
- Automatically set state to closed-loop when we tell the arm to move
- Safety check if closed-loop was aborted or other unexpected state is encountered before updating control state

### CODE REVIEW

**Review by cgiger00 — APPROVED** [review]

**Comment by cgiger00** [comment]
lgtm, cam probably has more input tho

File: jetson/ra_kinematics/mrover_arm.cpp [hunk 1]
```diff
@@ -209,10 +202,22 @@ void MRoverArm::target_orientation_callback(std::string channel, TargetOrientati
...
+    if (control_state != ControlState::WAITING_FOR_TARGET) {
+        std::cout << "IK calculations cancelled\n";
+
+        if (control_state != ControlState::OFF) {
+            std::cout << "UNEXPECTED BEHAVIOR: cancelled IK but did not leave closed-loop! Setting state to off.\n";
+            control_state = ControlState::OFF;
+        }
+
+        return;
+    }
```
Comment (CameronTressler) [inline jetson/ra_kinematics/mrover_arm.cpp:217]: Any way this could become a function as well?

**Comment by CameronTressler** [comment]
Didn't think of this till now but does it make sense to have OFF and WAITING_FOR_TARGET be different states?

File: jetson/ra_kinematics/mrover_arm.cpp [hunk 2]
```diff
@@ -502,6 +528,10 @@ void MRoverArm::zero_position_callback(std::string channel, Signal msg) {
 }
 
 void MRoverArm::arm_adjust_callback(std::string channel, ArmAdjustments msg) {
+    if (control_state == ControlState::OFF) {
+        set_to_closed_loop();
+    }
+
```
Comment (CameronTressler) [inline jetson/ra_kinematics/mrover_arm.cpp:534]: Honestly I don't think this is needed, and we can also remove the 4 lines right below this. We're about to call target_orientation, which should do both of these checks for us. If you disagree lmk

**Comment by bli76** [comment]
> Didn't think of this till now but does it make sense to have OFF and WAITING_FOR_TARGET be different states?

Guess not, especially now that we automatically switch to closed-loop if it was off.

**Comment by CameronTressler** [comment]
> > Didn't think of this till now but does it make sense to have OFF and WAITING_FOR_TARGET be different states?
> 
> Guess not, especially now that we automatically switch to closed-loop if it was off.

Okay I guess let’s leave as is for now, I don’t think it’s worth the potential to break something that isn’t broken

**Comment by CameronTressler** [comment]
Did you run 6 joint arm in the sim to test behavior of this since your last changes?

**Comment by bli76** [comment]
> Did you run 6 joint arm in the sim to test behavior of this since your last changes?

Yep, everything seems to work fine including shutting off closed-loop midway.

**Comment by bli76** [comment]
> > > Didn't think of this till now but does it make sense to have OFF and WAITING_FOR_TARGET be different states?
> > 
> > 
> > Guess not, especially now that we automatically switch to closed-loop if it was off.
> 
> Okay I guess let’s leave as is for now, I don’t think it’s worth the potential to break something that isn’t broken

Actually I took another look at the code and I think we need the separate states to know when to update the Closed Loop button on the GUI

**Comment by CameronTressler** [comment]
> > Did you run 6 joint arm in the sim to test behavior of this since your last changes?
> 
> Yep, everything seems to work fine including shutting off closed-loop midway.

Lit

**Comment by CameronTressler** [comment]
> > > Didn't think of this till now but does it make sense to have OFF and WAITING_FOR_TARGET be different states?
> > 
> > 
> > Guess not, especially now that we automatically switch to closed-loop if it was off.
> 
> Actually I took another look at the code and I think we need the separate states to know when to update the Closed Loop button on the GUI

Ah yeah you're right.

**Pull request merged**